### PR TITLE
Support Azure OpenAI API

### PR DESCRIPTION
**Pull request by sunshine0523**
Let openai-java support access Azure OpenAI API

### CODE REVIEW

**Comment by axiangcoding** [comment]
related to https://github.com/TheoKanning/openai-java/issues/249, we are looking forward to this pr being merged

**Comment by sunshine0523** [comment]
> related to #249, we are looking forward to this pr being merged

Yes, I saw this issue, but it seems that Azure OpenAI is not just about changing the baseUrl. Am I misunderstood?

**Comment by simonlmartin** [comment]
Also looking forward to this PR being merged... would it be possible to add a constructor which  enables the abiilty to pass in the AzureOpenAiApi class like the OpenAIService does so it is possible to add custom interceptors to the  chain?

**Review by derekhsu — APPROVED** [review]

**Review by iuadastra — APPROVED** [review]

**Comment by iuadastra** [comment]
@TheoKanning  can this be merged? It's been pending for too long

**Comment by wilmarDupont** [comment]
Looking forward to it. It would be incredibly helpful to me.

**Comment by zhaosen241** [comment]
Also looking forward to it, please merge it.

**Review by zhaosen241 — APPROVED** [review]

**Review by axiangcoding — APPROVED** [review]

File: service/src/main/java/com/theokanning/openai/service/OpenAiService.java [hunk 1]
```diff
@@ -4,10 +4,7 @@
 import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.PropertyNamingStrategy;
-import com.theokanning.openai.DeleteResult;
-import com.theokanning.openai.OpenAiApi;
-import com.theokanning.openai.OpenAiError;
-import com.theokanning.openai.OpenAiHttpException;
+import com.theokanning.openai.*;
```
Comment (bowenliang123) [inline service/src/main/java/com/theokanning/openai/service/OpenAiService.java:7]: Unnecessary changes to imports of `OpenAiService`.  Unrelated to the purposed feature. Changes should be reverted.

File: service/src/main/java/com/theokanning/openai/service/AzureOpenAiService.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        }
+    }
+
+    /**
+     * Calls the Open AI api and returns a Flowable of SSE for streaming
+     * omitting the last message.
+     *
+     * @param apiCall The api call
+     */
+    public static Flowable<SSE> stream(retrofit2.Call<ResponseBody> apiCall) {
```
Comment (bowenliang123) [inline service/src/main/java/com/theokanning/openai/service/AzureOpenAiService.java:181]:
```suggestion
    public static Flowable<SSE> stream(Call<ResponseBody> apiCall) {
```
Simplify it to `Call`, as `retrofit2.Call` is already imported.

File: service/src/main/java/com/theokanning/openai/service/AzureOpenAiService.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        return stream(apiCall, false);
+    }
+
+    /**
+     * Calls the Open AI api and returns a Flowable of SSE for streaming.
+     *
+     * @param apiCall  The api call
+     * @param emitDone If true the last message ([DONE]) is emitted
+     */
+    public static Flowable<SSE> stream(retrofit2.Call<ResponseBody> apiCall, boolean emitDone) {
```
Comment (bowenliang123) [inline service/src/main/java/com/theokanning/openai/service/AzureOpenAiService.java:191]: ditto

Comment (sunshine0523) [reply service/src/main/java/com/theokanning/openai/service/AzureOpenAiService.java:191]: I have also done other related work, but this pull request has not been merged, so I did not submit the new work

Comment (bowenliang123) [reply service/src/main/java/com/theokanning/openai/service/AzureOpenAiService.java:191]: What you mentioned sounds not a blocker here to simplify the reference here. Each PR should be standalone and good to reflect changes to the purpose.

File: client/src/main/java/com/theokanning/openai/AzureOpenAiApi.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import com.theokanning.openai.finetune.FineTuneResult;
+import com.theokanning.openai.model.Model;
+import io.reactivex.Single;
+import okhttp3.MultipartBody;
+import okhttp3.RequestBody;
+import okhttp3.ResponseBody;
+import retrofit2.Call;
+import retrofit2.http.*;
+
+public interface AzureOpenAiApi{
```
Comment (bowenliang123) [inline client/src/main/java/com/theokanning/openai/AzureOpenAiApi.java:20]:
```suggestion
public interface AzureOpenAiApi {
```

Format the code please.

**Comment by bowenliang123** [comment]
I am worried about the too much duplicated code to `OpenAiService` in this feature.
Not a good practice in software design patterns. Also bad in maintenance.
In this way, we will have many different similar versions with duplication for different vendors of Open AI services.

**Comment by sunshine0523** [comment]
> I am worried about the too much duplicated code to `OpenAiService` in this feature.
> Not a good practice in software design patterns. Also bad for future maintenance.
> In this way, we will have many different similar versions with duplication for different vendors of Open AI services.

I agree with you, I changed the library for my own demo. But now I've changed the implementation. The implementation is similar to OpenAI's Python library. I think that way is good.

**Comment by qixiaobo** [comment]
Maybe next release?

**Comment by Hakenadu** [comment]
I'd also love to see this merged, as otherwise we would have to implement our own client for this purpose.
Are there any plans for this?

**Comment by jnunderwood** [comment]
Is there a status update for this PR? We could really use it. Thank you!

**Comment by bowenliang123** [comment]
Any response or update to the review comments in the unnecessary code changes, the worries about duplicate code as existing `OpenAiService`? I don't think it's in a good enough shape.

**Comment by sunshine0523** [comment]
I think this code was written very poorly, so I closed this pull request

**Comment by bowenliang123** [comment]
At least the minor code suggestions should be good to be updated.
As `open-java` lib is shipped with Apache Kyuubi to provide ChatGPT as the chatting engine, we do concern about the quality and maintenance of the upstream dependencies.

**Comment by sunshine0523** [comment]
> At least the minor code suggestions should be good to be updated. As `open-java` lib is shipped with Apache Kyuubi to provide ChatGPT as the chatting engine, we do concern about the quality and maintenance of the upstream dependencies.

This project has not been updated for some time. Currently, there are other libraries that support Azure OpenAI. Perhaps you can consider replacing the active maintenance project.

**Comment by df007df** [comment]
> > At least the minor code suggestions should be good to be updated. As `open-java` lib is shipped with Apache Kyuubi to provide ChatGPT as the chatting engine, we do concern about the quality and maintenance of the upstream dependencies.
> 
> This project has not been updated for some time. Currently, there are other libraries that support Azure OpenAI. Perhaps you can consider replacing the active maintenance project.

你倒是说用哪个可以
